hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -251,15 +257,52 @@ public static HoodieIndex createUserDefinedIndex(HoodieWriteConfig config) {
    */
   private static <R> HoodieData<HoodieRecord<R>> getExistingRecords(
       HoodieData<Pair<String, String>> partitionLocations, HoodieWriteConfig config, HoodieTable hoodieTable) {
-    final Option<String> instantTime = hoodieTable
-        .getMetaClient()
+    HoodieTableMetaClient metaClient = hoodieTable.getMetaClient();
+    final Option<String> instantTime = metaClient
         .getActiveTimeline() // we need to include all actions and completed
         .filterCompletedInstants()
         .lastInstant()
         .map(HoodieInstant::requestedTime);
-    return partitionLocations.flatMap(p
-        -> new HoodieMergedReadHandle(config, instantTime, hoodieTable, Pair.of(p.getKey(), p.getValue()))
-        .getMergedRecords().iterator());
+    ReaderContextFactory<R> readerContextFactory = hoodieTable.getContext().getReaderContextFactory(metaClient);
+    if (instantTime.isEmpty()) {
+      return hoodieTable.getContext().emptyHoodieData();
+    }
+    return partitionLocations.flatMap(p -> {
+      Option<FileSlice> fileSliceOption = Option.fromJavaOptional(hoodieTable
+          .getHoodieView()
+          .getLatestMergedFileSlicesBeforeOrOn(p.getLeft(), instantTime.get())
+          .filter(fileSlice -> fileSlice.getFileId().equals(p.getRight()))
+          .findFirst());
+      if (fileSliceOption.isEmpty()) {
+        return Collections.emptyIterator();
+      }
+      Schema dataSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getWriteSchema()), config.allowOperationMetadataField());
+      Option<InternalSchema> internalSchemaOption = SerDeHelper.fromJson(config.getInternalSchema());
+      FileSlice fileSlice = fileSliceOption.get();
+      HoodieReaderContext<R> readerContext = readerContextFactory.getContext();
+      HoodieFileGroupReader<R> fileGroupReader = HoodieFileGroupReader.<R>newBuilder()
+          .withReaderContext(readerContext)
+          .withHoodieTableMetaClient(metaClient)
+          .withLatestCommitTime(instantTime.get())
+          .withFileSlice(fileSlice)
+          .withDataSchema(dataSchema)
+          .withRequestedSchema(dataSchema)
+          .withInternalSchema(internalSchemaOption)
+          .withShouldUseRecordPosition(false)
+          .withProps(metaClient.getTableConfig().getProps())
+          .build();
+      try {
+        final HoodieRecordLocation currentLocation = new HoodieRecordLocation(fileSlice.getBaseInstantTime(), fileSlice.getFileId());
+        return new CloseableMappingIterator<>(fileGroupReader.getClosableHoodieRecordIterator(), hoodieRecord -> {
+          hoodieRecord.unseal();
+          hoodieRecord.setCurrentLocation(currentLocation);
+          hoodieRecord.seal();
```

Comment:
Is it possible for the file group reader to return an iterator of `HoodieRecord`s that are already attached with the current location, since the file group reader knows the file slice, i.e, the base instant time and file ID?  Then the logic here can be simpler.

Reply:
I think we would still need the unseal and seal steps if I am not mistaken. If other parts of the code need HoodieRecords but do not require the location, then this could be extra work that those parts of the code incur if we move this to the FGReader code.

Reply:
+1 to @the-other-tim-brown , let's keep the FG reader cleaner and mainly aims for engine native record read support.

Reply:
Got it.  My intention is to add a new API that returns an iterator with the unseal and seal logic without affecting other callers.  This is fine for now since the logic is only one-off.